internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
+				ID:          batchID,
+				FinalizedAt: batchFinalizedAt,
+			})
+		})
+	}
+
+	// This could be written more simply using multiple `sliceutil.Map`s, but
+	// it's done this way to allocate as few new slices as necessary.
+	mapIDsAndFinalizedAt := func(setStateBatch map[int64]*batchCompleterSetState) ([]int64, []time.Time) {
+		var (
+			batchIDs         = make([]int64, len(setStateBatch))
+			batchFinalizedAt = make([]time.Time, len(setStateBatch))
+			i                int
+		)
+		for _, setState := range setStateBatch {
+			batchIDs[i] = setState.Params.ID
+			batchFinalizedAt[i] = *setState.Params.FinalizedAt
+			i++
+		}
+		return batchIDs, batchFinalizedAt
+	}
+
+	// Tease apart enormous batches into sub-batches.
+	//
+	// All the code below is concerned with doing that, with a fast loop that
+	// doesn't allocate any additional memory in case the entire batch is
+	// smaller than the sub-batch maximum size (which will be the common case).
+	const oneOperationMax = 2_000
+
+	var (
+		batchID, batchFinalizedAt = mapIDsAndFinalizedAt(setStateBatch)
+		jobRows                   []*rivertype.JobRow
+	)
+	if len(setStateBatch) > oneOperationMax {
+		jobRows = make([]*rivertype.JobRow, 0, len(setStateBatch))
+		for i := 0; i < len(setStateBatch); i += oneOperationMax {
+			endIndex := min(i+oneOperationMax, len(batchID)) // beginning of next sub-batch or end of slice
+			jobRowsSubBatch, err := insertSubBatch(batchID[i:endIndex], batchFinalizedAt[i:endIndex])
+			if err != nil {
+				return err
+			}
+			jobRows = append(jobRows, jobRowsSubBatch...)
+		}
+	} else {
+		var err error
+		jobRows, err = insertSubBatch(batchID, batchFinalizedAt)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, jobRow := range jobRows {
+		setState := setStateBatch[jobRow.ID]
+		setState.Stats.CompleteDuration = c.TimeNowUTC().Sub(setState.WaitingAt)
+		c.sendJobToSubscription(jobRow, setState.Stats)
+	}
+
 	return nil
 }
 
-func (c *AsyncJobCompleter) Subscribe(subscribeFunc func(update CompleterJobUpdated)) {
-	c.subscribeFuncMu.Lock()
-	defer c.subscribeFuncMu.Unlock()
+func (c *BatchCompleter) JobSetStateIfRunning(ctx context.Context, stats *jobstats.JobStatistics, params *riverdriver.JobSetStateIfRunningParams) error {
+	// Send completions other than setting to `complete` to an async completer.
+	// We consider this okay because these are expected to be much more rare, so
+	// only optimizing `complete` will yield huge speed gains.
+	if params.State != rivertype.JobStateCompleted {
+		return c.asyncCompleter.JobSetStateIfRunning(ctx, stats, params)
+	}
 
-	c.subscribeFunc = subscribeFunc
+	// Wait until the completer is started and ready to start processing
+	// batches. Alternatively, we could remove this and allow batches to start
+	// accumulating even if the service isn't started, but that could introduce
+	// some danger of the service never being started and therefore accumulating
+	// forever without completing jobs.
+	<-c.ready
```

Comment:
Can we instead make sure the completer is started & good to go before we even start working jobs? Seems better than having this channel receive on every single completion IMO. And the slight increase in startup time also seems worth it to be as sure as possible that jobs are ready to be worked successfully.

Reply:
Yep, changed that. And actually, we only bother to fully wait for startup in the benchmark tests, where it's most appropriate to mandate a fully ready completer.